doc/man-pages/man1/rxdebug.1
.\" Automatically generated by Pod::Man 2.16 (Pod::Simple 3.05)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sh \" Subsection heading
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.Sh), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.ie \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.el \{\
.    de IX
..
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "RXDEBUG 1"
.TH RXDEBUG 1 "2010-02-11" "OpenAFS" "AFS Command Reference"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
rxdebug \- Provides debugging trace of Rx activity
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBrxdebug\fR \fB\-servers\fR\ <\fIserver\ machine\fR> [\fB\-port\fR\ <\fI\s-1IP\s0\ port\fR>]
    [\fB\-nodally\fR] [\fB\-allconnections\fR] [\fB\-rxstats\fR] [\fB\-onlyserver\fR]
    [\fB\-onlyclient\fR] [\fB\-onlyport\fR\ <\fIshow\ only\ port\fR>]
    [\fB\-onlyhost\fR\ <\fIshow\ only\ host\fR>]
    [\fB\-onlyauth\fR\ <\fIshow\ only\ auth\ level\fR>] [\fB\-version\fR]
    [\fB\-noconns\fR] [\fB\-peers\fR] [\fB\-help\fR]
.PP
\&\fBrxdebug\fR \fB\-s\fR\ <\fIserver\ machine\fR> [\fB\-po\fR\ <\fI\s-1IP\s0\ port\fR>] [\fB\-nod\fR]
    [\fB\-a\fR] [\fB\-r\fR] [\fB\-onlys\fR] [\fB\-onlyc\fR] [\fB\-onlyp\fR\ <\fIshow\ only\ port\fR>]
    [\fB\-onlyh\fR\ <\fIshow\ only\ host\fR>] [\fB\-onlya\fR\ <\fIshow\ only\ auth\ level\fR>]
    [\fB\-v\fR] [\fB\-noc\fR] [\fB\-pe\fR] [\fB\-h\fR]
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
The \fBrxdebug\fR command provides a trace of Rx activity for the server or
client machine named by the \fB\-servers\fR argument. Rx is \s-1AFS\s0's proprietary
remote procedure call (\s-1RPC\s0) protocol, so this command enables the issuer
to check the status of communication between the Cache Manager or an \s-1AFS\s0
server process (as specified with the \fB\-port\fR argument) on the machine
and one or more processes on other machines.
.SH "OPTIONS"
.IX Header "OPTIONS"
.IP "\fB\-servers\fR <\fIserver machine\fR>" 4
.IX Item "-servers <server machine>"
Specifies the machine that is running the Cache Manager or server process
for which to trace Rx activity. Provide the machine's \s-1IP\s0 address in dotted
decimal format, its fully qualified host name (for example,
\&\f(CW\*(C`fs1.abc.com\*(C'\fR), or the shortest abbreviated form of its host name that
distinguishes it from other machines. Successful use of an abbreviated
form depends on the availability of a name resolution service (such as the
Domain Name Service or a local host table) at the time the command is
issued.
.IP "\fB\-port\fR <\fI\s-1IP\s0 port\fR>" 4
.IX Item "-port <IP port>"
Specifies the process for which to trace Rx activity. Omit this argument
to specify the File Server (\fBfileserver\fR process), or provide one of the
following values:
.RS 4
.ie n .IP "7000\fR for the File Server (\fBfileserver process)" 4
.el .IP "\f(CW7000\fR for the File Server (\fBfileserver\fR process)" 4
.IX Item "7000 for the File Server (fileserver process)"
.PD 0
.ie n .IP "7001 for the Cache Manager (specifically, its callback interface)" 4
.el .IP "\f(CW7001\fR for the Cache Manager (specifically, its callback interface)" 4
.IX Item "7001 for the Cache Manager (specifically, its callback interface)"
.ie n .IP "7002\fR for the Protection Server (\fBptserver process)" 4
.el .IP "\f(CW7002\fR for the Protection Server (\fBptserver\fR process)" 4
.IX Item "7002 for the Protection Server (ptserver process)"
.ie n .IP "7003\fR for the Volume Location (\s-1VL\s0) Server (\fBvlserver process)" 4
.el .IP "\f(CW7003\fR for the Volume Location (\s-1VL\s0) Server (\fBvlserver\fR process)" 4
.IX Item "7003 for the Volume Location (VL) Server (vlserver process)"

NEWS
OpenAFS News -- history of user-visible changes.

* Changes incorporated in OpenAFS 1.4.0

** Addition of pthreading support in volserver.

** Threading optimization of RX libraries.

** Windows 95/98 client desupported.

** Digital UNIX/Tru64 desupported.

** Windows XP/2003 client includes persistent (across boots) cache support.

** Windows XP/2003 now supports a cache of up to 1.2gb.

** Windows XP/2003 has various improvements to the included GUI and command 
   line tools.

** Windows XP/2003 now supports AFS UNC paths.

** MacOS X (10.3) and Linux use system inodes instead of a private inode pool.

** Files larger than 2gb are supported in Unix clients, and in servers if the 
   --enable-largefile-fileserver switch is enabled at configure time.

** Obsolete and insecure auxiliary network daemons (rcp, rsh, ftpd et al)
   have been removed.

** Fileserver optimizations have been made to improve timeout and callback 
   breaking behavior to inaccessible clients or clients whose address has
   changed.

** Kerberos 5 support is integrated.

** -nosettime is now the default for afsd.  Use "-settime" to get the
   old behavior.

** Mountpoint directory information is now only faked for cross-cell
   mountpoints when using the -fakestat flag (e.g. for the directories
   under /afs, but not for most other volumes mounted inside the cell).
   The -fakestat-all switch can be used to fake information for all
   mountpoints.

** When fakestat is enabled on MacOSX, the Finder can be used to browse
   a fully-populated /afs directory.  However, this precludes reliable
   use of entire volumes as MacOS bundles (i.e. containing a Contents
   directory in the root of the volume).

** Mountpoint directory information can be faked by the cache manager,
   making operations such as stat'ing all cells under /afs much faster.
   This is enabled by passing -fakestat to afsd, but might not be stable
   on all platforms.

* Changes incorporated in OpenAFS 1.2.9

** The kaserver now defaults to not allowing interrealm authentication,
   due to security vulnerabilities in the krb4 protocol.  The new
   "-crossrealm" flag to the kaserver is provided to reenable interrealm
   authentication if desired.

** RedHat Linux 9.0 is now supported.

** Solaris 9 12/02 is now supported.  Solaris 7 and 8 x86 should now
   work again.

** On Linux machines using 2.2 series kernels, 2.2.19 or higher is now
   required.

** An OpenAFS 1.2.9 afsd will not work with kernel modules built from
   an earlier OpenAFS release.  In general, using a mismatched afsd and
   kernel modules set is unsupported; it is not recommended that you use
   such a configuration on a regular basis.

* Changes incorporated in OpenAFS 1.2.8

** Mountpoint directory information is now only faked for cross-cell
   mountpoits when using the -fakestat flag (e.g. for the directories
   under /afs, but not for most other volumes mounted inside the cell).
   The -fakestat-all switch can be used to fake information for all
   mountpoints.

** HPUX 11.0 is now supported.

** It is now possible for AFS to use Kerberos 5 directly, via rxkad 2b.
   See the OpenAFS 1.2.8 Release Notes for more information on using this
   capability.

** An NFS translator kernel module is now included and compiled by default
   for Solaris only.

* Changes incorporated in OpenAFS 1.2.7

** MacOS X 10.2 is now supported.  FreeBSD 4.3 and later support is included

src/rxdebug/rxdebug.c
/*
 * Copyright 2000, International Business Machines Corporation and others.
 * All Rights Reserved.
 * 
 * This software has been released under the terms of the IBM Public
 * License.  For details, see the LICENSE file in the top-level source
 * directory or online at http://www.openafs.org/dl/license10.html
 */

#include <afsconfig.h>
#include <afs/param.h>


#include <sys/types.h>
#include <errno.h>
#ifdef AFS_NT40_ENV
#include <winsock2.h>
#else
#include <sys/socket.h>
#include <sys/file.h>
#include <netdb.h>
#include <arpa/inet.h>
#endif
#ifdef HAVE_NETINET_IN_H
#include <netinet/in.h>
#endif
#ifdef HAVE_SYS_TIME_H
#include <sys/time.h>
#endif
#ifdef HAVE_UNISTD_H
#include <unistd.h>
#endif
#include <string.h>
#include <sys/stat.h>
#include <afs/afsutil.h>
#include <afs/stds.h>
#include <afs/cmd.h>

#include <stdio.h>

#include <rx/rx_user.h>
#include <rx/rx_clock.h>
#include <rx/rx_queue.h>
#include <rx/rx.h>
#include <rx/rx_globals.h>


#define	TIMEOUT	    20

static short
PortNumber(char *aport)
{
    register int tc;
    register short total;

    total = 0;
    while ((tc = *aport++)) {
	if (tc < '0' || tc > '9')
	    return -1;		/* bad port number */
	total *= 10;
	total += tc - (int)'0';
    }
    return htons(total);
}

static short
PortName(char *aname)
{
    register struct servent *ts;
    ts = getservbyname(aname, NULL);
    if (!ts)
	return -1;
    return ts->s_port;		/* returns it in network byte order */
}

int
MainCommand(struct cmd_syndesc *as, void *arock)
{
    register int i;
    osi_socket s;
    int j;
    struct sockaddr_in taddr;
    afs_int32 host;
    struct in_addr hostAddr;
    short port;
    struct hostent *th;
    register afs_int32 code;
    int nodally;
    int allconns;
    int rxstats;
    int onlyClient, onlyServer;
    afs_int32 onlyHost;
    short onlyPort;
    int onlyAuth;
    int flag;
    int dallyCounter;
    int withSecStats;
    int withAllConn;
    int withRxStats;
    int withWaiters;
    int withIdleThreads;
    int withWaited;
    int withPeers;
    int withPackets;
    struct rx_debugStats tstats;
    char *portName, *hostName;
    char hoststr[20];
    struct rx_debugConn tconn;
    short noConns;
    short showPeers;
    short showLong;
    int version_flag;
    char version[64];
    afs_int32 length = 64;

    afs_uint32 supportedDebugValues = 0;
    afs_uint32 supportedStatValues = 0;
    afs_uint32 supportedConnValues = 0;
    afs_uint32 supportedPeerValues = 0;
    afs_int32 nextconn = 0;
    afs_int32 nextpeer = 0;

    nodally = (as->parms[2].items ? 1 : 0);
    allconns = (as->parms[3].items ? 1 : 0);
    rxstats = (as->parms[4].items ? 1 : 0);
    onlyServer = (as->parms[5].items ? 1 : 0);
    onlyClient = (as->parms[6].items ? 1 : 0);
    version_flag = (as->parms[10].items ? 1 : 0);
    noConns = (as->parms[11].items ? 1 : 0);
    showPeers = (as->parms[12].items ? 1 : 0);
    showLong = (as->parms[13].items ? 1 : 0);

    if (as->parms[0].items)
	hostName = as->parms[0].items->data;
    else
	hostName = NULL;

    if (as->parms[1].items)
	portName = as->parms[1].items->data;
    else
	portName = NULL;

    if (as->parms[7].items) {
	char *name = as->parms[7].items->data;
	if ((onlyPort = PortNumber(name)) == -1)
	    onlyPort = PortName(name);
	if (onlyPort == -1) {
	    printf("rxdebug: can't resolve port name %s\n", name);
	    exit(1);
	}
    } else
	onlyPort = -1;

    if (as->parms[8].items) {
	char *name = as->parms[8].items->data;
	struct hostent *th;
	th = hostutil_GetHostByName(name);
	if (!th) {
	    printf("rxdebug: host %s not found in host table\n", name);
	    exit(1);
	}
	memcpy(&onlyHost, th->h_addr, sizeof(afs_int32));
    } else
	onlyHost = -1;

    if (as->parms[9].items) {
	char *name = as->parms[9].items->data;
	if (strcmp(name, "clear") == 0)
	    onlyAuth = 0;
	else if (strcmp(name, "auth") == 0)
	    onlyAuth = 1;
	else if (strcmp(name, "crypt") == 0)
	    onlyAuth = 2;
	else if ((strcmp(name, "null") == 0) || (strcmp(name, "none") == 0)
		 || (strncmp(name, "noauth", 6) == 0)
		 || (strncmp(name, "unauth", 6) == 0))
	    onlyAuth = -1;
	else {
	    fprintf(stderr, "Unknown authentication level: %s\n", name);
	    exit(1);
	}
    } else
	onlyAuth = 999;

    /* lookup host */
    if (hostName) {
	th = hostutil_GetHostByName(hostName);
	if (!th) {
	    printf("rxdebug: host %s not found in host table\n", hostName);
	    exit(1);
	}
	memcpy(&host, th->h_addr, sizeof(afs_int32));
    } else
	host = htonl(0x7f000001);	/* IP localhost */

    if (!portName)
	port = htons(7000);	/* default is fileserver */
    else {
	if ((port = PortNumber(portName)) == -1)
	    port = PortName(portName);
	if (port == -1) {
	    printf("rxdebug: can't resolve port name %s\n", portName);
	    exit(1);
	}
    }

    dallyCounter = 0;

    hostAddr.s_addr = host;
    afs_inet_ntoa_r(hostAddr.s_addr, hoststr);
    printf("Trying %s (port %d):\n", hoststr, ntohs(port));
    s = socket(AF_INET, SOCK_DGRAM, 0);
    taddr.sin_family = AF_INET;
    taddr.sin_port = 0;
    taddr.sin_addr.s_addr = 0;
#ifdef STRUCT_SOCKADDR_HAS_SA_LEN
    taddr.sin_len = sizeof(struct sockaddr_in);
#endif
    code = bind(s, (struct sockaddr *)&taddr, sizeof(struct sockaddr_in));
    if (code) {
	perror("bind");
	exit(1);
    }

    if (version_flag) {

	code = rx_GetServerVersion(s, host, port, length, version);
	if (code < 0) {
	    printf("get version call failed with code %d, errno %d\n", code,
		   errno);
	    exit(1);
	}
	printf("AFS version: %s\n", version);
	fflush(stdout);

	exit(0);

    }


    code = rx_GetServerDebug(s, host, port, &tstats, &supportedDebugValues);
    if (code < 0) {
	printf("getstats call failed with code %d\n", code);
	exit(1);
    }

    withSecStats = (supportedDebugValues & RX_SERVER_DEBUG_SEC_STATS);
    withAllConn = (supportedDebugValues & RX_SERVER_DEBUG_ALL_CONN);
    withRxStats = (supportedDebugValues & RX_SERVER_DEBUG_RX_STATS);
    withWaiters = (supportedDebugValues & RX_SERVER_DEBUG_WAITER_CNT);
    withIdleThreads = (supportedDebugValues & RX_SERVER_DEBUG_IDLE_THREADS);
    withWaited = (supportedDebugValues & RX_SERVER_DEBUG_WAITED_CNT);
    withPeers = (supportedDebugValues & RX_SERVER_DEBUG_ALL_PEER);
    withPackets = (supportedDebugValues & RX_SERVER_DEBUG_PACKETS_CNT);

    if (withPackets)
        printf("Free packets: %d/%d, packet reclaims: %d, calls: %d, used FDs: %d\n",
               tstats.nFreePackets, tstats.nPackets, tstats.packetReclaims, 
               tstats.callsExecuted, tstats.usedFDs);
    else
        printf("Free packets: %d, packet reclaims: %d, calls: %d, used FDs: %d\n",
               tstats.nFreePackets, tstats.packetReclaims, tstats.callsExecuted,
               tstats.usedFDs);
    if (!tstats.waitingForPackets)
	printf("not ");
    printf("waiting for packets.\n");
    if (withWaiters)
	printf("%d calls waiting for a thread\n", tstats.nWaiting);
    if (withIdleThreads)
	printf("%d threads are idle\n", tstats.idleThreads);
    if (withWaited)
	printf("%d calls have waited for a thread\n", tstats.nWaited);

    if (rxstats) {
	if (!withRxStats) {
	  noRxStats:
	    withRxStats = 0;
	    fprintf(stderr,
		    "WARNING: Server doesn't support retrieval of Rx statistics\n");
	} else {
	    union {
		struct rx_statistics rxstats;
		struct rx_debugIn debug;
	    } packet;

	    /* should gracefully handle the case where rx_stats grows */
	    code =
		rx_GetServerStats(s, host, port, &packet.rxstats,
				  &supportedStatValues);
	    if (code < 0) {
		printf("rxstats call failed with code %d\n", code);
		exit(1);
	    }
	    if (code != sizeof(packet.rxstats)) {
		if (packet.debug.type == RX_DEBUGI_BADTYPE)
		    goto noRxStats;
		printf
		    ("WARNING: returned Rx statistics of unexpected size (got %d)\n",
		     code);
		/* handle other versions?... */
	    }

	    rx_PrintTheseStats(stdout, &packet.rxstats, sizeof(packet.rxstats),
			       tstats.nFreePackets, tstats.version);
	}
    }

    if (!noConns) {
	if (allconns) {
	    if (!withAllConn)
		fprintf(stderr,
			"WARNING: Server doesn't support retrieval of all connections,\n         getting only interesting instead.\n");
	}

	if (onlyServer)
	    printf("Showing only server connections\n");
	if (onlyClient)
	    printf("Showing only client connections\n");
	if (onlyAuth != 999) {
	    static char *name[] =
		{ "unauthenticated", "rxkad_clear", "rxkad_auth",
		"rxkad_crypt"
	    };
	    printf("Showing only %s connections\n", name[onlyAuth + 1]);
	}
	if (onlyHost != -1) {
	    hostAddr.s_addr = onlyHost;
	    afs_inet_ntoa_r(hostAddr.s_addr, hoststr);
	    printf("Showing only connections from host %s\n",
		   hoststr);
	}
	if (onlyPort != -1)
	    printf("Showing only connections on port %u\n", ntohs(onlyPort));

	for (i = 0;; i++) {
	    code =
		rx_GetServerConnections(s, host, port, &nextconn, allconns,
					supportedDebugValues, &tconn,
					&supportedConnValues);
	    if (code < 0) {
		printf("getconn call failed with code %d\n", code);
		break;
	    }
	    if (tconn.cid == (afs_int32) 0xffffffff) {
		printf("Done.\n");
		break;
	    }

	    /* see if we're in nodally mode and all calls are dallying */
	    if (nodally) {
		flag = 0;
		for (j = 0; j < RX_MAXCALLS; j++) {
		    if (tconn.callState[j] != RX_STATE_NOTINIT
			&& tconn.callState[j] != RX_STATE_DALLY) {
			flag = 1;
			break;
		    }
		}
		if (flag == 0) {
		    /* this call looks too ordinary, bump skipped count and go
		     * around again */
		    dallyCounter++;
		    continue;
		}
	    }
	    if ((onlyHost != -1) && (onlyHost != tconn.host))
		continue;
	    if ((onlyPort != -1) && (onlyPort != tconn.port))
		continue;
	    if (onlyServer && (tconn.type != RX_SERVER_CONNECTION))
		continue;
	    if (onlyClient && (tconn.type != RX_CLIENT_CONNECTION))
		continue;
	    if (onlyAuth != 999) {
		if (onlyAuth == -1) {
		    if (tconn.securityIndex != 0)
			continue;
		} else {
		    if (tconn.securityIndex != 2)
			continue;
		    if (withSecStats && (tconn.secStats.type == 3)
			&& (tconn.secStats.level != onlyAuth))
			continue;
		}
	    }

	    /* now display the connection */
	    hostAddr.s_addr = tconn.host;
	    afs_inet_ntoa_r(hostAddr.s_addr, hoststr);
	    printf("Connection from host %s, port %hu, ", hoststr,
		   ntohs(tconn.port));
	    if (tconn.epoch)
		printf("Cuid %x/%x", tconn.epoch, tconn.cid);
	    else
		printf("cid %x", tconn.cid);
	    if (tconn.error)
		printf(", error %d", tconn.error);
	    printf("\n  serial %d, ", tconn.serial);
	    printf(" natMTU %d, ", tconn.natMTU);

	    if (tconn.flags) {
		printf("flags");
		if (tconn.flags & RX_CONN_MAKECALL_WAITING)
		    printf(" MAKECALL_WAITING");
		if (tconn.flags & RX_CONN_DESTROY_ME)
		    printf(" DESTROYED");
		if (tconn.flags & RX_CONN_USING_PACKET_CKSUM)
		    printf(" pktCksum");
                if (tconn.flags & RX_CONN_KNOW_WINDOW)
                    printf(" knowWindow");
                if (tconn.flags & RX_CONN_RESET)
                    printf(" reset");
                if (tconn.flags & RX_CONN_BUSY)
                    printf(" busy");
                if (tconn.flags & RX_CONN_ATTACHWAIT)
                    printf(" attachWait");
		printf(", ");
	    }
	    printf("security index %d, ", tconn.securityIndex);
	    if (tconn.type == RX_CLIENT_CONNECTION)
		printf("client conn\n");
	    else
		printf("server conn\n");

	    if (withSecStats) {
		switch ((int)tconn.secStats.type) {
		case 0:
		    if (tconn.securityIndex == 2)
			printf
			    ("  no GetStats procedure for security object\n");
		    break;
		case 1:
		    printf("  rxnull level=%d, flags=%d\n",
			   tconn.secStats.level, tconn.secStats.flags);
		    break;
		case 2:
		    printf("  rxvab level=%d, flags=%d\n",
			   tconn.secStats.level, tconn.secStats.flags);
		    break;
		case 3:{
			char *level;
			char flags = tconn.secStats.flags;
			if (tconn.secStats.level == 0)
			    level = "clear";
			else if (tconn.secStats.level == 1)
			    level = "auth";
			else if (tconn.secStats.level == 2)
			    level = "crypt";
			else
			    level = "unknown";
			printf("  rxkad: level %s", level);
			if (flags)
			    printf(", flags");
			if (flags & 1)
			    printf(" unalloc");
			if (flags & 2)
			    printf(" authenticated");
			if (flags & 4)
			    printf(" expired");
			if (flags & 8)
			    printf(" pktCksum");
			if (tconn.secStats.expires)
			    /* Apparently due to a bug in the RT compiler that
			     * prevents (afs_uint32)0xffffffff => (double) from working,
			     * this code produces negative lifetimes when run on the
			     * RT. */
			    printf(", expires in %.1f hours",
				   ((afs_uint32) tconn.secStats.expires -
				    time(0)) / 3600.0);
			if (!(flags & 1)) {
			    printf("\n  Received %u bytes in %u packets\n",
				   tconn.secStats.bytesReceived,
				   tconn.secStats.packetsReceived);
			    printf("  Sent %u bytes in %u packets\n",
				   tconn.secStats.bytesSent,
				   tconn.secStats.packetsSent);
			} else
			    printf("\n");
			break;
		    }

		default:
		    printf("  unknown\n");
		}
	    }

	    for (j = 0; j < RX_MAXCALLS; j++) {
		printf("    call %d: # %d, state ", j, tconn.callNumber[j]);
		if (tconn.callState[j] == RX_STATE_NOTINIT) {
		    printf("not initialized\n");
		    continue;
		} else if (tconn.callState[j] == RX_STATE_PRECALL)
		    printf("precall, ");
		else if (tconn.callState[j] == RX_STATE_ACTIVE)
		    printf("active, ");
		else if (tconn.callState[j] == RX_STATE_DALLY)
		    printf("dally, ");
		else if (tconn.callState[j] == RX_STATE_HOLD)
		    printf("hold, ");
		printf("mode: ");
		if (tconn.callMode[j] == RX_MODE_SENDING)
		    printf("sending");
		else if (tconn.callMode[j] == RX_MODE_RECEIVING)
		    printf("receiving");
		else if (tconn.callMode[j] == RX_MODE_ERROR)
		    printf("error");
		else if (tconn.callMode[j] == RX_MODE_EOF)
		    printf("eof");
		else
		    printf("unknown");
		if (tconn.callFlags[j]) {
		    printf(", flags:");
		    if (tconn.callFlags[j] & RX_CALL_READER_WAIT)
			printf(" reader_wait");
		    if (tconn.callFlags[j] & RX_CALL_WAIT_WINDOW_ALLOC)
			printf(" window_alloc");
		    if (tconn.callFlags[j] & RX_CALL_WAIT_WINDOW_SEND)
			printf(" window_send");
		    if (tconn.callFlags[j] & RX_CALL_WAIT_PACKETS)
			printf(" wait_packets");
		    if (tconn.callFlags[j] & RX_CALL_WAIT_PROC)
			printf(" waiting_for_process");
		    if (tconn.callFlags[j] & RX_CALL_RECEIVE_DONE)
			printf(" receive_done");
		    if (tconn.callFlags[j] & RX_CALL_CLEARED)
			printf(" call_cleared");
		}
		if (tconn.callOther[j] & RX_OTHER_IN)
		    printf(", has_input_packets");
		if (tconn.callOther[j] & RX_OTHER_OUT)
		    printf(", has_output_packets");
		printf("\n");
	    }
	}
	if (nodally)
	    printf("Skipped %d dallying connections.\n", dallyCounter);
    }
    if (showPeers && withPeers) {
	for (i = 0;; i++) {
	    struct rx_debugPeer tpeer;
	    code =
		rx_GetServerPeers(s, host, port, &nextpeer, allconns, &tpeer,
				  &supportedPeerValues);
	    if (code < 0) {
		printf("getpeer call failed with code %d\n", code);
		break;
	    }
	    if (tpeer.host == 0xffffffff) {
		printf("Done.\n");
		break;
	    }

	    if ((onlyHost != -1) && (onlyHost != tpeer.host))
		continue;
	    if ((onlyPort != -1) && (onlyPort != tpeer.port))
		continue;

	    /* now display the peer */
	    hostAddr.s_addr = tpeer.host;
	    afs_inet_ntoa_r(hostAddr.s_addr, hoststr);
	    printf("Peer at host %s, port %hu\n", hoststr, 
		   ntohs(tpeer.port));
	    printf("\tifMTU %hu\tnatMTU %hu\tmaxMTU %hu\n", tpeer.ifMTU,
		   tpeer.natMTU, tpeer.maxMTU);
	    printf("\tpackets sent %u\tpacket resends %u\n", tpeer.nSent,
		   tpeer.reSends);
	    printf("\tbytes sent high %u low %u\n", tpeer.bytesSent.high,
		   tpeer.bytesSent.low);
	    printf("\tbytes received high %u low %u\n",
		   tpeer.bytesReceived.high, tpeer.bytesReceived.low);
	    printf("\trtt %u msec, rtt_dev %u msec\n", tpeer.rtt >> 3,
		   tpeer.rtt_dev >> 2);
	    printf("\ttimeout %u.%03u sec\n", tpeer.timeout.sec,
		   tpeer.timeout.usec / 1000);
	    if (!showLong)
		continue;

	    printf("\tin/out packet skew: %d/%d\n", tpeer.inPacketSkew,
		    tpeer.outPacketSkew);
	    printf("\tcongestion window %d, MTU %d\n", tpeer.cwind,
		   tpeer.MTU);
	    printf("\tcurrent/if/max jumbogram size: %d/%d/%d\n",
		   tpeer.nDgramPackets, tpeer.ifDgramPackets,
		   tpeer.maxDgramPackets);
	}
    }
    exit(0);
}

/* simple main program */
#ifndef AFS_NT40_ENV
#include "AFS_component_version_number.c"
#endif
int
main(int argc, char **argv)
{
    struct cmd_syndesc *ts;

#ifdef RXDEBUG
    rxi_DebugInit();
#endif
#ifdef AFS_NT40_ENV
    if (afs_winsockInit() < 0) {
	printf("%s: Couldn't initialize winsock. Exiting...\n", argv[0]);
	return 1;
    }
#endif

    ts = cmd_CreateSyntax(NULL, MainCommand, NULL, "probe RX server");
    cmd_AddParm(ts, "-servers", CMD_SINGLE, CMD_REQUIRED, "server machine");
    cmd_AddParm(ts, "-port", CMD_SINGLE, CMD_OPTIONAL, "IP port");
    cmd_AddParm(ts, "-nodally", CMD_FLAG, CMD_OPTIONAL,
		"don't show dallying conns");
    cmd_AddParm(ts, "-allconnections", CMD_FLAG, CMD_OPTIONAL,
		"don't filter out uninteresting connections on server");
    cmd_AddParm(ts, "-rxstats", CMD_FLAG, CMD_OPTIONAL, "show Rx statistics");
    cmd_AddParm(ts, "-onlyserver", CMD_FLAG, CMD_OPTIONAL,
		"only show server conns");
    cmd_AddParm(ts, "-onlyclient", CMD_FLAG, CMD_OPTIONAL,
		"only show client conns");
    cmd_AddParm(ts, "-onlyport", CMD_SINGLE, CMD_OPTIONAL,
		"show only <port>");
    cmd_AddParm(ts, "-onlyhost", CMD_SINGLE, CMD_OPTIONAL,
		"show only <host>");
    cmd_AddParm(ts, "-onlyauth", CMD_SINGLE, CMD_OPTIONAL,
		"show only <auth level>");

    cmd_AddParm(ts, "-version", CMD_FLAG, CMD_OPTIONAL,
		"show AFS version id");
    cmd_AddParm(ts, "-noconns", CMD_FLAG, CMD_OPTIONAL,
		"show no connections");
    cmd_AddParm(ts, "-peers", CMD_FLAG, CMD_OPTIONAL, "show peers");
    cmd_AddParm(ts, "-long", CMD_FLAG, CMD_OPTIONAL, "detailed output");

    cmd_Dispatch(argc, argv);
    exit(0);
}
